WebSocket support

rubenseyer opened the pull request:
I too dream of a Mumble on the web, so here's my attempt at giving Grumble built-in Websocket support. It's not quite ready for prime time, and would need a full-featured client to match, but I still think we can discuss the architecture. I'm a total Go noob, so any suggestions on the code are welcome as well. Sorry for also sticking another unrelated bug I found in here as well, but I didn't think a missing exclamation mark warranted it's own pull request.

(I'm using [this client](https://github.com/Johni0702/mumble-web) to test, although it isn't feature complete.)

Websockets and the Mumble protocol are exposed on the same port, since conveniently they both run on top of TLS. Protocol suggestions through NPN/ALPN are followed, but since neither the Mumble client nor some browsers (looking at you, Chrome) perform this negotiation the code looks at the first two bytes to see if they match "GE", otherwise we assume the Mumble protocol. There is no practical difference in the code for clients speaking the raw Mumble protocol. Websocket clients are upgraded and then wrapped into a mock connection which packs and unpacks messages transparently.

Even certificates work through the web, if the clients are set up correctly. The HTTP server is set to request but not force certificates. This causes Firefox to pop up a certificate selection UI, but sadly [Chrome devs prefer not to](https://bugs.chromium.org/p/chromium/issues/detail?id=329884) (it does work if you trick it to cache a certificate for the host, but it's not very user-friendly). It's very useful to have a valid certificate for the server because the browser can otherwise silently refuse to connect.

Websockets are TCP-only and additionally has their own framing which may mess with the packets and affect performance. However, as long as the client is fairly cooperative you can tunnel voice over them with acceptable quality, although the browser really kills the performance if the tab isn't in focus. To really have full web voice support it should ideally also communicate over WebRTC (which would be UDP), but that is more complex to implement and requires an additional negotiation on the application layer which would mean a Mumble protocol extension.

mkrautz commented:
Thank you. I have wanted WebSocket support in Murmur/Grumble for a long time -- but there are too many other issues that take up my time to be able to work on it. But I have thought about it quite a bit.

Just a few drive-by comments. Haven't done a thorough review yet:

For it to *really* work, I think we'd need *good* Let's Encrypt/ACME integration. Probably with the `x/crypto/acme/autocert` package. But that can always be added later. :-)

I'm happy that you've tested against https://github.com/Johni0702/mumble-web. That's what I would have suggested you test against.

The negotiation/sniffer is probably too simplistic. Also, Chrome doesn't negotiate http or h2 via ALPN? Aww. That's sad.

I've always wanted to expose the WebSocket on the native Mumble port like you've done, but perhaps we should expose it via port 443 as well. Perhaps with a protocol extension allowing you to pick the "real" port. (This is future work.)

Regarding client certs. Does Firefox bring up the picker when you connect to the WebSocket? I actually hoped that it wouldn't. I think that it's best not to use client certs directly for WebSockets, and instead move client cert handling into the application layer instead of TLS. (Mostly to avoid "bad" browser UI for client certs). Browsers are pretty clear that client certs are not something they like, so trying to force use of them via TLS is not going to work in the long term. (Not requesting a client certificate obviously clashes with the ability for WebSocket support to use the same port. So perhaps the rule should be that port 443 does not request a client cert, while native Mumble ports do.)

Hope to be able to do a full review sooner rather than later! Thanks again.

mkrautz commented:
> The negotiation/sniffer is probably too simplistic. Also, Chrome doesn't negotiate http or h2 via ALPN? Aww. That's sad.

Actually, if it's just for the WS upgrade, I suppose it's fine? I was thinking it had to catch all HTTP verbs, but reading the code, that's probably not the case?

rubenseyer commented:
Here comes a very long response... but discussing how the implementation would actually work in practice was what I wanted to do so here goes:

> The negotiation/sniffer is probably too simplistic. [...]
>
> Actually, if it's just for the WS upgrade, I suppose it's fine? I was thinking it had to catch all HTTP verbs, but reading the code, that's probably not the case?

It's more of a design decision here, I think. The WS spec says that the upgrade request must be a GET. After those two bytes are sniffed (if necessary) and the code decides to treat it as a WS connection everything is handled through the HTTP and WS implementations - there's no detection on every packet or something like that. I see it this way: if you try to use other HTTP methods, i.e. definitely not upgrade to WS, the result would be exactly the same as trying to talk HTTP to a server only speaking the Mumble protocol. It's like any Mumble server by default unless you "speak the magic words". If you want to handle every HTTP method on the initial request through the HTTP server (the design decision I referred to earlier) the current sniffer is not enough. I supposed all of this is moot if we're adding port 443 because of client cert issues anyway.

> Also, Chrome doesn't negotiate http or h2 via ALPN? Aww. That's sad.

As for Chrome, for some reason they completely clear the ALPN protocol list when sending WS requests. (I thought it was a bug at first so I even tracked down the [exact source line](https://cs.chromium.org/chromium/src/net/http/http_stream_factory_impl_job.cc?rcl=c9f54f7b9990ac0eb0b961c7433c846cdc1e94c3&l=976) if you're curious.) No idea why they do that whatsoever. They play nice like everyone else for ordinary HTTP. Implementing HTTP/2 to force a negotiation there would be meaningless as there is no standard for WS over HTTP/2 anyway.

> Regarding client certs. Does Firefox bring up the picker when you connect to the WebSocket? I actually hoped that it wouldn't. I think that it's best not to use client certs directly for WebSockets, and instead move client cert handling into the application layer instead of TLS. (Mostly to avoid "bad" browser UI for client certs). Browsers are pretty clear that client certs are not something they like, so trying to force use of them via TLS is not going to work in the long term.

Firefox does bring up the certificate UI (given that you have a cert, see below). The Chrome devs in the bug report I linked earlier speculated that this isn't deliberate and more of a coincidence, and as you say, client certs are not a priority feature on the web. I don't think decoupling client certs from the TLS connection over WS is too difficult, seeing as there really only are two spots in the code that retrieve the cert directly from the connection - the new client handling and the user information request.

> (Not requesting a client certificate obviously clashes with the ability for WebSocket support to use the same port. So perhaps the rule should be that port 443 does not request a client cert, while native Mumble ports do.)

The shared port that requests a client cert could have been a good compromise, if only browsers would cooperate. If you don't have any client certs Firefox just doesn't send one and continues - works just like you'd hope to. For those few who actually have a cert installed they can click Cancel once and then they will never see it again. However, with Chrome, even if you have no certs, the WS connection just fails unless a decision already has been made for ordinary HTTP. Therefore, just browsing to the web client and trying to connect will simply fail. The aforementioned Chrome bug report said this is working as intended. This means that everything would only flow nicely if Grumble also served the client, which really isn't what I had in mind. (Also, I realised now while testing this again that the extra [TLSConfig I put in the HTTP server](https://github.com/mumble-voip/grumble/pull/18/commits/def71812fd29c2527d6d10456d225251092fa2a5#diff-ce993bb3f4a68decdeae018ec068ec3aR40) to make sure certs were sent from browsers is entirely useless - I'll change this together with any other suggested changes after you've done a full review.)

> Hope to be able to do a full review sooner rather than later!

Take your time. I am also exploring patches to the browser clients to keep me busy 😄 .

rubenseyer commented:
Alright, so I thought about this a bit more and made some changes to the code after thinking about your comments. Hopefully you hadn't gotten too far with the review, because I probably threw out two-thirds of the code.

- I dropped WS on port 64738 entirely for simplicity. Seeing as you don't care for the browser client cert ui, it would only work otherwise in specific browsers, and we're going to have to listen on port 443 anyway for the Let's Encrypt integration, I can find no use case for it (although if you _really_ wanted it I could add it back)
- Instead I moved out a "real" HTTP server on port 443. It still upgrades WebSocket connections just like before, but is no longer hidden inside `handleIncomingClient` (it was a nifty way to avoid adding more state to the server and slot the server under the native TLS listener, but it really was a hack). I guess it would make some API etc. possible without too much pain if one wanted that too, but that would be some (far) future work.
- Obviously, this breaks client certificates for WS on the TLS level (they're disabled on the port 443 listener). Doesn't matter if this is going to be handled in the application layer.
- I'm not a fan of having to run the server privileged (at least on Unix-based systems, I don't think this is a problem on Windows?) to be able to bind to port 443, so some sort of privilege dropping after bind would need to be added.
- The Travis CI build seems to have broken because the graceful shutdown of the HTTP server didn't yet exist in that Go version.

Again, take your time. I don't expect this to get merged until it's done for real, and that probably means adding at least the Let's Encrypt integration and the application layer client certs for WS.

mkrautz commented:
FYI, I just bumped the .travis.yml to use Go 1.9.  If you rebase against master, I hope the graceful http shutdown build failure should be fixed.

rubenseyer commented:
All good. Thanks!

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+
+var upgrader = websocket.Upgrader{
+	HandshakeTimeout: 20 * time.Second,
+	Subprotocols:     []string{"mumble1.2.5", "binary"},
```
Should this be 1.2.4? IIRC, there were no protocol changes between 1.2.4 and 1.2.5.

mkrautz replied:
Preferably just "mumble". If we break compat, we'd just do "mumble2", "mumble3", etc. (or whatever)

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+	"time"
+)
+
+type WrappedSocketListener struct {
```
WrappedSocketListener cannot be reused after being closed, correct?  If so, add a `closed int32` field to this struct to signal if Close has already been called, and if close the done channel.

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func (wsl *WrappedSocketListener) Accept() (net.Conn, error) {
+	select {
```
Dont let function be called after closed. E.g.:

    if atomic.LoadInt32(&wsl.Closed) { return nil, errors.New("already closed") }

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func (wsl *WrappedSocketListener) Close() error {
+	wsl.done <- struct{}{}
```
```
if atomic.CompareAndSwap(&wsl.closed, 0, 1) {
    return errors.New("already closed")
}
close(wsl.done)
return nil

bontibon commented on pkg/web/websocket.go:
```diff
@@ -0,0 +1,66 @@
+
+type WrappedSocket struct {
+	ws        *websocket.Conn
+	msgbuf    *bytes.Buffer
```
Change to a non-pointer;  the zero value of `bytes.Buffer` is ready to use.

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+			logger.Printf("%v", err)
+			return
+		}
+		wsl.sockets <- &WrappedSocket{ws: conn, msgbuf: new(bytes.Buffer)}
```
don't initialize `msgbuf`; the zero value is OK to use.

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+	addr    net.Addr
+}
+
+func (wsl *WrappedSocketListener) Accept() (net.Conn, error) {
```
personal preference, but change `wsl` to a single letter (e.g. `w`); I find multi-letter receiver variable names messy.

bontibon commented on pkg/web/websocket.go:
```diff
@@ -0,0 +1,66 @@
+	msgbuf    *bytes.Buffer
+}
+
+func (wws *WrappedSocket) Read(b []byte) (n int, err error) {
```
single letter receiver variable name again

bontibon commented on pkg/web/websocket.go:
```diff
@@ -0,0 +1,66 @@
+	return wws.ws.SetWriteDeadline(t)
+}
+
+func (wws *WrappedSocket) UnderlyingWebSocket() *websocket.Conn {
```
Is this method used anywhere? Could just omit it until it's needed.

bontibon commented on pkg/web/http.go:
```diff
@@ -0,0 +1,69 @@
+	},
+}
+
+func NewHTTPServer(addr net.Addr, tlscfg *tls.Config, logger *log.Logger) (*http.Server, net.Listener) {
```
Returning an `*http.Server` seems wrong here; it should be the caller's responsibility to create and manage the server.  I think it makes more sense to return a concrete type that implements an `http.Handler` and `net.Listener`.

rubenseyer commented:
Thanks again. (I know this is a lot of commits. I'll squash after you're done.)

rubenseyer commented on pkg/web/wslisten.go:
```diff
@@ -43,6 +43,20 @@ func (w *WrappedSocketListener) Addr() net.Addr {
+		http.Error(rw, http.StatusText(http.StatusServiceUnavailable), http.StatusServiceUnavailable)
+		return
+	}
+	w.logger.Printf("Upgrading web connection from: %v", r.RemoteAddr)
```
An alternative to keeping a pointer to the logger might be to get it like this:
```go
r.Context().Value(http.ServerContextKey).(*http.Server).ErrorLog
```
(and handle if it's nil). I don't know which of these you prefer.

bontibon commented on pkg/web/wslisten.go:
```diff
@@ -0,0 +1,75 @@
+	"github.com/gorilla/websocket"
+)
+
+type WrappedSocketListener struct {
```
We're already in the package `web`, so we can simplify this name.  `Listener` might be a good choice.

bontibon commented on pkg/web/wslisten.go:
```diff
@@ -0,0 +1,75 @@
+	},
+}
+
+func ListenWS(laddr net.Addr, logger *log.Logger) *WrappedSocketListener {
```
Rename this function to `NewListener`.

bontibon commented on pkg/web/websocket.go:
```diff
@@ -0,0 +1,63 @@
+	"github.com/gorilla/websocket"
+)
+
+type WrappedSocket struct {
```
This type doesn't need to be exported, so let's hide it;  rename to `conn`.

rubenseyer commented:
Thanks!

mkrautz commented on cmd/grumble/server.go:
```diff
@@ -1344,7 +1352,17 @@ func (server *Server) Port() int {
 
-// Returns the port the server is currently listning
+// Returns the port the web server will listen on when it is
+// started. Returns 0 on failure.
```
> Returns 0 on failure.
Doesn't seem like it?

rubenseyer replied:
Line 1338 says it too (I just implemented the web port retrieval in exactly the same way). I'll get rid of the comment for both.

rubenseyer commented:
Rebased everything onto the new master. If you'd like me to start on the Let's Encrypt integration, I want to know how you think it should be configured. Config file? Command-line flags? Interactive prompts? At a minimum, there has to be a mechanism for TOS acceptance (and probably for domain name input).

mkrautz commented on pkg/web/wslisten.go:
```diff
@@ -0,0 +1,75 @@
+	l.sockets <- &conn{ws: ws}
+}
+
+var upgrader = websocket.Upgrader{
```
I'd move this further up the file. Probably below the imports.

mkrautz commented on pkg/web/wslisten.go:
```diff
@@ -0,0 +1,75 @@
+	},
+}
+
+func NewListener(laddr net.Addr, logger *log.Logger) *Listener {
```
I'd move NewListener just below the definition of type Listener.

mkrautz commented on cmd/grumble/server.go:
```diff
@@ -1461,7 +1504,12 @@ func (server *Server) Stop() (err error) {
-	// Close the TLS listener and the TCP listener
+	// Wait for the HTTP server to shutdown gracefully
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(15*time.Second))
+	server.webhttp.Shutdown(ctx)
```
Missing error handling.

mkrautz commented on cmd/grumble/server.go:
```diff
@@ -1461,7 +1504,12 @@ func (server *Server) Stop() (err error) {
 
-	// Close the TLS listener and the TCP listener
+	// Wait for the HTTP server to shutdown gracefully
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(15*time.Second))
```
Why the 15 second timeout? Did you encounter a problem? Why not just `Shutdown(ctx.Background())`?
I don't mind it, but it needs a reason -- and if there is a reason, it needs a comment.

mkrautz commented on cmd/grumble/server.go:
```diff
@@ -1461,7 +1504,12 @@ func (server *Server) Stop() (err error) {
+	// Wait for the HTTP server to shutdown gracefully
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(15*time.Second))
+	server.webhttp.Shutdown(ctx)
+	cancel()
```
Why cancel after Shutdown has returned?

bontibon replied:
This is recommended behaviour by the context package. It's done in order to prevent the timer from leaking (you'll also get a `go vet` error if you don't call the cancel function).

mkrautz replied:
Thanks, I've not done much Go after contexts were introduced.

mkrautz commented on cmd/grumble/server.go:
```diff
@@ -256,27 +263,28 @@ func (server *Server) handleIncomingClient(conn net.Conn) (err error) {
-		client.Disconnect()
-		return
-	}
+	if tlsconn, ok := client.conn.(*tls.Conn); ok {
```
Please add a comment describing why. Maybe "Only consider client certificates for direct connections, not WebSocket connections. We do not support TLS-level client certificates for WebSocket client."

mkrautz commented on cmd/grumble/message.go:
```diff
@@ -1328,7 +1328,7 @@ func (server *Server) handleUserStatsMessage(client *Client, msg *Message) {
 
 	if details {
-		if tlsconn := target.conn.(*tls.Conn); tlsconn != nil {
+		if tlsconn, ok := target.conn.(*tls.Conn); ok {
```
I'd add a comment about not supporting client certificates for WebSocket connections here, too.

mkrautz commented:
> Rebased everything onto the new master. If you'd like me to start on the Let's Encrypt integration, I want to know how you think it should be configured. Config file? Command-line flags? Interactive prompts? At a minimum, there has to be a mechanism for TOS acceptance (and probably for domain name input).

That'd be very nice!

Not sure. I think an interactive "first run" experience would be nice, but I don't think it's a requirement.

I think a config file option would suffice.
It should probably be blank by default. If it is set, you've agreed to the Let's Encrypt TOS.
We can explain this in a comment in the config file, I'd think.

...But I just realized that we don't really have a config file for Grumble anymore.

Feel free to suggest something, or just do a JSON config file. I'd probably pre-process the JSON, reading each line, if a line contains "//" or "#", disregard the those characters and the rest of the line. That way we can have comments in there. Open to other suggestions.

I'd think the config file should probably just be called config.json (or whatever format we choose) and live in the data dir.

bontibon commented on pkg/web/wslisten.go:
```diff
@@ -0,0 +1,75 @@
+	return l.addr
+}
+
+func (l *Listener) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
```
Convention is to have `w` as the variable name for `http.ResponseWriter`.

mkrautz commented:
Regarding the config file: maybe, to avoid too much trouble, grumble should have a murmur emulation mode. So it'd be possible to use a regular murmur.ini with grumble.

That doesn't preclude us from doing something else alongside that, but it'd probably help a lot of people out. It'd be neat if grumble could just be a drop-in replacement for murmur in many scenarios.

rubenseyer commented:
Thanks. These changes should do it. If there really is nothing else, just tell me when I can squash the WebSocket changes (or maybe if you don't care about the unrelated 19d8d35 just do it through the github ui). I think a new pull request for the config stuff afterwards would be better (but there's no code yet, so...)

Regarding everything in `Stop()`; it's never actually called from anywhere... (And it currently doesn't work anyway due to the double close of listeners causing it to return, but that bug is probably out of scope for this pull request.) I still thought it important to put the HTTP server shutdown there so it isn't forgotten in case the server stop will be used someday.

> I'd probably pre-process the JSON, reading each line, if a line contains "//" or "#", disregard the those characters and the rest of the line. That way we can have comments in there.
> Regarding the config file: maybe, to avoid too much trouble, grumble should have a murmur emulation mode. So it'd be possible to use a regular murmur.ini with grumble.

I am unsure about rolling our own JSON-with-comments parser. I'd prefer just picking a format with comments in the spec instead, although JSON has the upside that it would not require an external dependency. However, if we're parsing ini-like formats anyway (because the murmur.ini compatibility is indeed useful) I think we might as well go with config.ini.

mkrautz commented:
@rubenseyer Feel free to do the squash, keeping the separate commit separate.

mkrautz commented:
Once again: thanks a bunch!
